NVDA objects/UIA: introduce UIA drag and drop event handlers

josephsl opened the pull request:
Hi,
To be marked ready for review once NVDA 2021.1 beta 1 is released and master branch moves onto 2021.2:

### Link to issue number:
Closes #12271 

### Summary of the issue:
UIA drag and drop events are used to communicate that items were rearranged on screen or dragged around. This is used when:

* Rearranging Start menu tiles
* Rearranging Action Center quick items
* In Windows Insider Preview Cobalt build 21337 and later, after reordering virtual desktops

### Description of how this pull request fixes the issue:
Introduce UIA drag complete and drop target dropped event handlers so NVDA can announce at least the dragged item and its new positoin. The event handler can be expanded to add support for vision enhancement providers and other goddies upon request.

### Testing strategy:
Tested through Windows 10 App Essentials add-on:

1. Rearranging Start menu tiles and observing that NVDA announces old and new tile positions.
2. Rearranging Action Center items to make sure the new location was announced, both after rearranging items and opening Action Center afterwards.
3. In build 21337 and later, giving virtual desktops different names and reordering them.

### Known issues with pull request:
None

### Change log entries:
New features:

NVDA can announce new location of items such as Start menu tiles and Action Center items after items are rearranged. (#12271)

Possible changes for developers:

Added event handlers for UIA drag complete and drop target dropped events. (#12271)

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review and confirm you have considered the following items.
Mark items you have considered by checking them.
You can do this when editing the Pull request description with an x: `[ ]` becomes `[x]`.
You can also check the checkboxes after the PR is created.
-->

- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual tests.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.

Thanks.

josephsl committed:
UIA handler: add UIA drag complete and drop target dropped event Id's. Re #12271.

UIA drag complete and drop target dropped events are used to announce new location for an element. The former is seen when rearranging tiles in Windows 8 start screen, and the latter is triggered when arranging quick actions in Windows 10 Action Center. Although both events might as well call gain focus event handler, specific events will be defined in UIA objects for future-proofing.

josephsl committed:
NVDAObjects/UIA: define UIA drag complete event. Re #12271.

When tiles are rearranged in Windows 8 start screen, braille reports new location whereas speech doesn't. Therefore use UIA drag complete event to force focus to be reported.

josephsl committed:
NvDAObjects/UIA: define UIA drop target dropped event. Re #12271.

Unlike UIA drag complete event, drop target dropped event is raised by a containing element where items are being rearragned, the best example being rearranging quick actions in Windows 10 Action Center. At least NVDA does know what the focused element is, therefore raise a gain focus event on the focused object so the newly dragged item can be announced.

josephsl committed:
NVDAObjects/UIA: update copyright year

josephsl committed:
Merge branch 'nvaccess:master' into i12271UIADragDropEvents

josephsl committed:
UIA handler: mark UIA drag complete/drop target dropped event definitions with Flake8 F405 NOQA (definition). Re #12271

leonardder commented:
> To be marked ready for review once NVDA 2021.1 beta 1 is released and master branch moves onto 2021.2:

That has happened.

leonardder commented on source/_UIAHandler.py:
```diff
@@ -189,6 +189,8 @@
 	#UIA_ToolTipClosedEventId:"hide",
 	UIA_Window_WindowOpenedEventId:"UIA_window_windowOpen",
 	UIA_SystemAlertEventId:"UIA_systemAlert",
+	UIA_Drag_DragCompleteEventId: "UIA_dragComplete",  # NOQA: F405
```
Please avoid noqa

```suggestion
	UIA.UIA_Drag_DragCompleteEventId: "UIA_dragComplete",
```

josephsl replied:
NOQA is required because without that, local lint test fails- Flake8 doesn't know where these UIA event constants are coming from since we import everything from UIA client support library (COM interfaces).

leonardder replied:
Note that in my suggestion, I suggested to use the UIA import. That should avoid the Flake8 error.

leonardder commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -1832,6 +1832,24 @@ def event_UIA_notification(self, notificationKind=None, notificationProcessing=U
+		A prominent example is rearranging Start screen tiles on Windows 8 and later.
+		"""
+		# #12271: speech does not report new location but braille does.
+		self.reportFocus()
```
This puzzles me.

1. Why are you not mimicking gainFocus behavior completely here? Especially for vision, it makes sense to communicate the new location so the highlighter will follow.

josephsl replied:
Oops - looks like a mismatch between this PR and WinTenApps where the latter doees queue a gain focus event. I'll correct this today.

leonardder commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -1832,6 +1832,24 @@ def event_UIA_notification(self, notificationKind=None, notificationProcessing=U
+		"""
+		# #12271: unlike drag complete event, it is a different element that raises this event.
+		# but NVDA does record the newly dragged element as focused object.
+		api.getFocusObject().reportFocus()
```
What if the dragged element is entirely unrelated to the focused object?

josephsl replied:
That's a distinct possibility. We may need to find more places where UIA drag/drop events are fired - so far, prominent places I found are ones I noted in code comments.

LeonarddeR reviewed (changes requested):
To be honest, I think we should put some more thought into how drag and drop should behave in a screen reader before come up with the actual implementation.

leonardder commented:
Note that there's also the locationChange event. Would it be possible to use that instead?

josephsl commented:
hi,

Location change event is defined in IAccessible/internal WinEvent handler, therefore the approach in this PR to handle UIA events. I think we might as well create another PR to let base NVDA objects handle location changes before applying it to various accessibility API handlers. I'm looking into the possibility that UIA might have an equivalent event to location change MSAA event.

Thanks,

josephsl committed:
Merge branch 'master' into i12271UIADragDropEvents

josephsl committed:
NVDAObjects/UIA: replace obj.reportFocus with obj.event_gainFocus. Re #12271.

Pointed out by Leonard de Ruijter: obj.reportFocus does not allow vision enhancmenet providers and other components from reacting to gain focus event. Therefore call obj.event_gainFocus when drag complete and/or drop target dropped events are fired. Also, add a source code comment on drop target dropped event that points out which element fires this event.

josephsl commented:
Hi, I’ll look into it this weekend. Thanks.

josephsl committed:
Merge branch 'master' into i12271UIADragDropEvents

josephsl committed:
UIA handler/events: reference UIA module when defining event constants (integers). Re #12271.

Reminder from Leonard de Ruijter: reference UIA module when defining UIA event constants (UIA.constant).

seanbudd reviewed (commented):
Thanks @josephsl , I've tested the rearrangement of Windows 10 start menu tiles and this seems to work well. I've left a couple clarifying comments but otherwise LGTM.

leonardder commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -1832,6 +1832,25 @@ def event_UIA_notification(self, notificationKind=None, notificationProcessing=U
+		Raised by an element when a drag operation on it is completed.
+		A prominent example is rearranging Start screen tiles on Windows 8 and later.
+		"""
+		# #12271: speech does not report new location but braille does.
```
i do not understand this comment.

josephsl replied:
When I first implemented it (back in November 2019 under Windows 10 App Essentials add-on), I noticed that speech didn't report the newly dragged element but braille did. We may need to find out why and come up with replacements.

leonardder commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -1832,6 +1832,25 @@ def event_UIA_notification(self, notificationKind=None, notificationProcessing=U
+		"""
+		# #12271: unlike drag complete event, it is a different element that raises this event.
+		# Specifically, it is the containing element that raises this event.
+		# but NVDA does record the newly dragged element as focused object.
```
How is the newly dragged element focused by NVDA?

josephsl replied:
I think the following traceback may explain this a bit (debug log with MSAA and UIA flags turned on, along with Windows 10 App Essentials enabled; scenario: an item in Action Center was rearranged):

IO - inputCore.InputManager.executeGesture (01:23:37.906) - winInputHook (1140):
Input: kb(desktop):alt+shift+leftArrow
DEBUG - IAccessibleHandler.internalWinEventHandler.winEventCallback (01:23:37.927) - MainThread (7948):
Hook received winEvent: EVENT_OBJECT_FOCUS, window 3281400 (Windows.UI.Core.CoreWindow), objectID 176, childID 0, process 2844 (shellexperiencehost), thread 10584
DEBUG - IAccessibleHandler.internalWinEventHandler.winEventCallback (01:23:37.928) - MainThread (7948):
Adding winEvent to limiter: EVENT_OBJECT_FOCUS, window 3281400 (Windows.UI.Core.CoreWindow), objectID 176, childID 0, process 2844 (shellexperiencehost), thread 10584
DEBUG - IAccessibleHandler.internalWinEventHandler.winEventCallback (01:23:37.941) - MainThread (7948):
Hook received winEvent: EVENT_OBJECT_FOCUS, window 3281400 (Windows.UI.Core.CoreWindow), objectID 177, childID 0, process 2844 (shellexperiencehost), thread 10584
DEBUG - IAccessibleHandler.internalWinEventHandler.winEventCallback (01:23:37.941) - MainThread (7948):
Adding winEvent to limiter: EVENT_OBJECT_FOCUS, window 3281400 (Windows.UI.Core.CoreWindow), objectID 177, childID 0, process 2844 (shellexperiencehost), thread 10584
DEBUG - IAccessibleHandler.orderedWinEventLimiter.OrderedWinEventLimiter.flushEvents (01:23:37.946) - MainThread (7948):
Emitting winEvent EVENT_OBJECT_FOCUS, window 3281400 (Windows.UI.Core.CoreWindow), objectID 176, childID 0, process 2844 (shellexperiencehost), thread 10584
DEBUG - IAccessibleHandler.orderedWinEventLimiter.OrderedWinEventLimiter.flushEvents (01:23:37.946) - MainThread (7948):
Emitting winEvent EVENT_OBJECT_FOCUS, window 3281400 (Windows.UI.Core.CoreWindow), objectID 177, childID 0, process 2844 (shellexperiencehost), thread 10584
DEBUG - IAccessibleHandler.processFocusWinEvent (01:23:37.949) - MainThread (7948):
Processing focus winEvent: window 3281400 (Windows.UI.Core.CoreWindow), objectID 177, childID 0, process 2844 (shellexperiencehost), force False
DEBUG - IAccessibleHandler.winEventToNVDAEvent (01:23:37.949) - MainThread (7948):
Creating NVDA event from winEvent: EVENT_OBJECT_FOCUS, window 3281400 (Windows.UI.Core.CoreWindow), objectID 177, childID 0, process 2844 (shellexperiencehost), use cache False
DEBUG - IAccessibleHandler.winEventToNVDAEvent (01:23:37.950) - MainThread (7948):
winEvent mapped to NVDA event: gainFocus
DEBUG - IAccessibleHandler.winEventToNVDAEvent (01:23:37.950) - MainThread (7948):
Native UIA window. Dropping winEvent EVENT_OBJECT_FOCUS, window 3281400 (Windows.UI.Core.CoreWindow), objectID 177, childID 0, process 2844 (shellexperiencehost)
DEBUG - IAccessibleHandler.processFocusWinEvent (01:23:37.951) - MainThread (7948):
Processing focus winEvent: window 3281400 (Windows.UI.Core.CoreWindow), objectID 176, childID 0, process 2844 (shellexperiencehost), force False
DEBUG - IAccessibleHandler.winEventToNVDAEvent (01:23:37.951) - MainThread (7948):
Creating NVDA event from winEvent: EVENT_OBJECT_FOCUS, window 3281400 (Windows.UI.Core.CoreWindow), objectID 176, childID 0, process 2844 (shellexperiencehost), use cache False
DEBUG - IAccessibleHandler.winEventToNVDAEvent (01:23:37.951) - MainThread (7948):
winEvent mapped to NVDA event: gainFocus
DEBUG - IAccessibleHandler.winEventToNVDAEvent (01:23:37.951) - MainThread (7948):
Native UIA window. Dropping winEvent EVENT_OBJECT_FOCUS, window 3281400 (Windows.UI.Core.CoreWindow), objectID 176, childID 0, process 2844 (shellexperiencehost)
DEBUG - _UIAHandler.UIAHandler.IUIAutomationFocusChangedEventHandler_HandleFocusChangedEvent (01:23:37.982) - Dummy-3 (12344):
HandleFocusChangedEvent: Ignoring because no object or ignored by object itself
DEBUGWARNING - _UIAHandler.UIAHandler.IUIAutomationFocusChangedEventHandler_HandleFocusChangedEvent (01:23:37.985) - Dummy-3 (12344):
HandleFocusChangedEvent: Ignoring duplicate focus event
DEBUG - external:globalPlugins.wintenObjs.GlobalPlugin.uiaDebugLogging (01:23:38.021) - MainThread (7948):
W10: UIA object: <NVDAObjects.UIA.UIA object at 0x072874D0>, name: , event: dropTargetDropped, app module: <'shellexperiencehost' (appName 'shellexperiencehost', process ID 2844) at address 6cb1df0>, Automation Id: ScrollWrapper, class name: ScrollViewer

NOte that this was with selective UIA event registration enabled.

Thanks.

LeonarddeR reviewed (changes requested):
I'm still a bit reluctant to give this a go.

1. I don't think it is very good practice to use event_gainFocus in cases where it isn't triggered by a real gainFocus event that's coming through NVDA"s event handler.
2. I still tink it would be best to take this back to the design table to come up with a description of expected user experience.

josephsl commented:
Hi, I think both events will not be raised at the same time (I need to find an example that does this, but so far I can’t find it), and drop target dropped is fired when an element is dropped (will take a look at comment suggestion while looking into Leonard’s comments). Thanks.

josephsl commented:
Hi,

The resulting experience is that NVDA would announce the newly dragged element, similar in experience to how Narrator does it.

Thanks.

josephsl commented:
Hi,

I'm willing to revisit this PR throughout this summer, but since Leonard asked that we go back to the drawing board, I think it would be best to close it for now unless someone wants to work on it further based on what we have so far.

Thanks.

seanbudd commented:
Hi @josephsl, I think it's a good idea to leave these open for now and label them as needing a new author. Several of these NV Access may be able to pick up if another doesn't get to it first.

josephsl commented:
Hi, or perhaps let’s label it as “good for new dev” for people wanting to learn more about internals of UI Automation. Thanks.

leonardder commented:
I'm happy to take this one as soon as we agree about UX.

josephsl commented:
Hi, I guess I can say it now: when in doubt, learn from what Narrator does and does not do. By the way, according to UIA docs, drag and drop target events are available from Windows 8. Thanks.

josephsl commented:
Hi, by the way, whichever UX decision is chosen, I’ll apply it to WinTenApps add-on. Thanks.

XLTechie commented:
This is interesting and desirable work @josephsl, I for one would hate to 
see it get closed.

Although might this be a bit much for a new dev, especially given the high 
level disagreement about direction? Just a thought.

mazen428 commented:
Hello, In my opinion if at least NVDA could announce that an item is draggable would be great like narrator, right now you'd have to press NVDA F 1 and look at the bottom of the log to know which isn't really the best.
In Windows 11, the taskbar pinned tiles are draggable now, not sure if that was in Windows 10 too.

cary-rowen commented:
I don’t want to see this PR being abandoned. Does it have the latest news?

lukaszgo1 commented:
@cary-rowen No it does not as it has been clearly labeled as closed/needs-new-author which means that unless someone has time and motivation to continue this work it is stalled. @seanbudd  Perhaps PR's labeled as 'closed/needs-new-author' should be closed or at a minimum marked as drafts to avoid confusion?

josephsl commented:
Hi, specific reason: life priorities and PR design. In the meantime Windows App Essentials add-on provides this functionality, something that is really an alternative at this point (I hope someone can make this better soon).

seanbudd commented:
@lukaszgo1 - agreed, closing this PR so it can be taken over.

josephsl commented:
2022 update: see josephsl/wintenapps#75 as to why this PR should be taken care of sooner than later, as without this, users must rely on Windows App Essentials add-on to rearrange tiles and folders in Windows 11 Version 22H2 start menu.
